[4/5 verity read-only root] Add the verity-read-only-root package and integrate with imager tools

dmcilvaney opened the pull request:
<!--
COMMENT BLOCKS WILL NOT BE INCLUDED IN THE PR.
Feel free to delete sections of the template which do not apply to your PR, or add additional details
-->

###### Merge Checklist  <!-- REQUIRED -->
<!-- You can set them now ([x]) or set them later using the Github UI -->
**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`)
- ~~[ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)~~ Done in final PR
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- ~~[ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass~~ Partial changes won't pass testing, need all 5/5 PRs.
- ~~[ ] Documentation has been updated to match any changes to the build system~~ Documentation to follow later.
- [ ] Ready to merge


---

###### Summary <!-- REQUIRED -->
<!-- Quick explanation of the changes. -->
Read-only root and dm-verity: This set of 5 PRs will enable dm-verity based read-only root file systems which can detect, and optionally repair modifications to files on the root file system. This supports both basic resilience to corruption as well as a number of additional security features.

PR [4/5] adds the read-only-root dracut module which is responsible for mounting the verity disks. Integrates with the imager tool.

###### Change Log  <!-- REQUIRED -->
<!-- Detail the changes made here. -->
<!-- Please list any packages which will be affected by this change, if applicable. -->
<!-- Please list any CVES fixed by this change, if applicable. -->
- Add the `verity-read-only-root` package
  - This adds a new, optional `dracut` module which is able to parse and mount dm-verity roots
  - Updates the `initramfs` package to watch for this new package and update as needed
  - Enable the `CONFIG_DM_VERITY_FEC` flag for the kernel (FEC = Forward Error Correction)
- Create a read-only-root config field
  - Add example config file `read-only-root-efi.json`
  - Add `[ReadOnlyVerityRoot]` field to the json schema
    - Will auto-populate with sane defaults if `Enable=true` is set with everything else omitted.
- Add package checks to the previously created config package validation step.
  - Check kernel is set correctly
  - Check verity package is included if the ReadOnlyVerityRoot config is enabled
- `imager` tool will mount the root as a device-mapper device, then measure it via dm-verity once the image build is done. This needs to be the final step since any subsequent changes will invalidate the measurement.
  - Switch the root to read-only once done creating the image
  - Measure the image, optionally create error correcting data
  - Place the measurements and FEC data into the intitramfs
  - ** In theory this should work for the ISO installer as well, however there is a bug with the initramfs repackaging currently. It sees all files as symlinks during ISO install and never actually copies the files back into the archive.**
- Add `miscdebug` package with `WaitForUser()` and `WaitForDebugger()` functions.
  - Since we run as root and inside chroots it can be extremely challenging to break in with a debugger. Add two library functions to help break in via busyloop.

###### Does this affect the toolchain?  <!-- REQUIRED -->
<!-- Any packages which are included in the toolchain should be carefully considered. Make sure the toolchain builds with these changes if so. -->
NO


###### Test Methodology
Local builds and testing via hyper-v

MateuszMalisz commented on SPECS/verity-read-only-root/create_linear_debug_mount.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
nit: is it of any significance that some of the scripts invoke bash while others invoke sh?

dmcilvaney replied:
Yes, the scripts that run in the initramfs (not sure if by necessity or convention) all run /bin/sh, while the module-setup.sh which runs in the context of the full OS uses bin/bash.

---
In reply to: [552693738](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552693738) [](ancestors = 552693738)

MateuszMalisz commented on toolkit/imageconfigs/read-only-root-efi.json:
```diff
@@ -0,0 +1,101 @@
+            "Users": [
+                {
+                    "Name": "root",
+                    "Password": "p@ssw0rd"
```
Should this user have an explicitly set password?

schmittjoseph replied:
+1 I think we're trying to avoid the hard coded passwords here, but not sure if that's an option.

dmcilvaney replied:
Open to suggestions here. Unless you do something to setup a mount point or symlink (and I think shadow won't work with symlinks...) the user is also read-only.

I could switch to something SSH key related maybe?

---
In reply to: [552875082](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552875082) [](ancestors = 552875082)

dmcilvaney replied:
I've switched to a hashed password which is a better example to set, but I've left the password in a comment along with a mention of the ssh key. I think this is a good compromise (Jon was ok with the password some time ago when I talked with him about it).

---
In reply to: [552939716](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552939716) [](ancestors = 552939716,552875082)

MateuszMalisz commented on toolkit/tools/internal/miscdebug/miscdebug.go:
```diff
@@ -0,0 +1,38 @@
+
+	for i != 0 {
+		logger.Log.Errorf("Waiting for debugger %d, once broken in run `so` to step out, `set i=0`, then `c`", i)
+		ic(&i)
```
Is it simply `i++`? In that case I'd get rid of the helper function.

dmcilvaney replied:
Go will optimize too much if you use i++, and you can't break out anymore. This is some really hacky debugging garbage, best not to think too deeply on it :P

---
In reply to: [552713370](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552713370) [](ancestors = 552713370)

MateuszMalisz replied:
In that case it would be great to have a comment here mentioning that this function is left for debugging purposes.

MateuszMalisz commented on toolkit/tools/internal/miscdebug/miscdebug.go:
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package miscdebug
```
Why not e.g. `debugutils` instead of `miscdebug`?

MateuszMalisz commented on toolkit/tools/imagegen/installutils/installutils.go:
```diff
@@ -1523,6 +1535,54 @@ func setGrubCfgIMA(grubPath string, kernelCommandline configuration.KernelComman
+	logger.Log.Debugf("Adding Verity Root ('%s') to %s", verityArgs, grubPath)
+	err = sed(verityPattern, verityArgs, cmdline.GetSedDelimeter(), grubPath)
+	if err != nil {
+		logger.Log.Warnf("Failed to set grub.cfg's IMA setting: %v", err)
```
Shouldn't this force a build break?

dmcilvaney replied:
It should bubble the error out, this is the pattern we use on the other config update functions.

---
In reply to: [552717537](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552717537) [](ancestors = 552717537)

MateuszMalisz commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,316 @@
+	stdout, stderr, err := shell.Execute("dmsetup", "remove", v.MappedName)
+	if err != nil {
+		err = fmt.Errorf("Unable to clean up device mapper device '%s' '%s': %w", stdout, stderr, err)
+		return
```
You did not declare the error as a return value, hence you're not returning it here.

dmcilvaney replied:
Good catch, I'm going to both print this and return it since this would generally be a defered function

---
In reply to: [552719407](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552719407) [](ancestors = 552719407)

dmcilvaney replied:
As an aside, we need to go through and handle failures in defer calls.
```
defer func() {
    if deferError := deferedFunction(); deferError != nil {
        err = deferError
    }
}
```
This will set the out functions err value. I've noticed we don't handle unmounts reliably sometimes and never report or handle it.

---
In reply to: [552952854](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r552952854) [](ancestors = 552952854,552719407)

MateuszMalisz commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,316 @@
+	logger.Log.Info("Verifying the verity partition")
+	verityOutput, stderr, err = shell.Execute("veritysetup", verityVerifyArgs...)
+	if err != nil {
+		logger.Log.Error(verityOutput)
```
I suggest wrapping it in some context with `Errorf` to keep the message meaningful in the future

MateuszMalisz commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,316 @@
+// AddRootVerityFilesToInitramfs adds files needed for a verity root to the initramfs
+// - workingFolder is a temporary folder to extract the initramfs to
+// - initramfsPath is the path to the initramfs
+//func AddRootVerityFilesToInitramfs(destinationPath, initramfsDevice string, readOnlyDevice configuration.ReadOnlyVerityRoot) (err error) {
```
Seems like a stale comment

MateuszMalisz commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,316 @@
+	debugMountPoint     = "/mnt/verity_overlay_debug_tmpfs"
+)
+
+//VerityDevice represents a device mapper linear device used for a dm-verity read-only partition.
```
nit:space after `//`?

schmittjoseph commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,317 @@
+// - initramfsPath is the path to the initramfs
+func (v *VerityDevice) AddRootVerityFilesToInitramfs(workingFolder, initramfsPath string) (err error) {
+	var (
+		verityWorkingDirectory = filepath.Join(workingFolder, v.MappedName)
```
Nit, since assignment is happening here, use the `:=` variable initialization syntax.

```golang
verityWorkingDirectory := filepath.Join(workingFolder, v.MappedName)
````

schmittjoseph commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,317 @@
+	}
+
+	// Searches for a line like: "Root hash:      1234567890abcdefg..."
+	rootHashLine, stderr, err := shell.ExecuteWithStdin(verityOutput, "grep", "^Root hash:")
```
I think we should avoid calling grep for regex matching here. In general we use golang's built in regex engine. This avoids external dependencies and also provides the beneifits that come with calling into typed native code.

dmcilvaney replied:
Fair point, this was translated from a shell script. The regex actually simplifies this as well.

---
In reply to: [553013709](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r553013709) [](ancestors = 553013709)

schmittjoseph commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,317 @@
+	}
+	// Print the last entry in the line
+	// ORS (output record separator) inserts a newline normally, explicitly skip it
+	rootHash, stderr, err := shell.ExecuteWithStdin(rootHashLine, "awk", `BEGIN {ORS=""}; END{ print $NF }`)
```
See above comment, this could be handled with golang's built in regex engine.

schmittjoseph commented on toolkit/tools/imagegen/diskutils/verity.go:
```diff
@@ -0,0 +1,317 @@
+// - devicePath is the device to check
+func IsReadOnlyDevice(devicePath string) (result bool) {
+	verityPrefix := filepath.Join(mappingFilePath, mappingVerityPrefix)
+	if strings.HasPrefix(devicePath, verityPrefix) {
```
nit: just return `strings.HasPrefix(....)`.

schmittjoseph commented on toolkit/tools/imager/imager.go:
```diff
@@ -464,15 +484,42 @@ func buildImage(mountPointMap, mountPointToFsTypeMap, mountPointToMountArgsMap m
+		if systemConfig.ReadOnlyVerityRoot.Enable {
+			logger.Log.Info("Configuring root for read only with dm-verity")
+			var initramfsPathList []string
+			//runtime.Breakpoint()
```
leftover breakpoint?

dmcilvaney commented:
Rebased onto the final staging branch since they are good to go (only change from rebases was the removal of a single parenthesis in a comment).

Also changed the password in the config file.

christopherco commented on toolkit/tools/imagegen/configuration/veritydisk.go:
```diff
@@ -0,0 +1,128 @@
+// - RootHashSignatureEnable: Validate the root hash against a key stored in the kernel's
+//     system keyring. The signature file should be called "<Name>.p7" and must be stored in
+//     the initramfs. This signature WILL NOT BE included automatically in the initramfs. It must
+//     be included via an out of band build step.
```
Do we have an example of the out-of-band build step required for injecting the <Name>.p7 file into the initramfs?

christopherco commented on toolkit/tools/imagegen/installutils/installutils.go:
```diff
@@ -1529,6 +1541,54 @@ func setGrubCfgIMA(grubPath string, kernelCommandline configuration.KernelComman
+	logger.Log.Debugf("Adding Verity Root ('%s') to %s", verityArgs, grubPath)
+	err = sed(verityPattern, verityArgs, cmdline.GetSedDelimeter(), grubPath)
+	if err != nil {
+		logger.Log.Warnf("Failed to set grub.cfg's IMA setting: %v", err)
```
Warning says "IMA setting". Probably a typo. Should be ReadOnlyVerityRoot setting.

dmcilvaney commented on SPECS/kernel/kernel.spec:
```diff
@@ -3,7 +3,7 @@
 Name:           kernel
 Version:        5.4.83
-Release:        1%{?dist}
+Release:        2%{?dist}
```
That is done by a commit later on (b8c721). I haven't squashed any of the cleanup commits here yet.

---
In reply to: [555463976](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r555463976) [](ancestors = 555463976)

dmcilvaney commented on SPECS/verity-read-only-root/20verity-mount/verity-mount.sh:
```diff
@@ -0,0 +1,195 @@
+    if [ -n ${verity_overlays} ]; then
+        # Create working directories for overlays
+        mkdir -p "${OVERLAY_TMPFS}"
+        mount -t tmpfs tmpfs -o ${OVERLAY_MNT_OPTS},size=20% "${OVERLAY_TMPFS}" || \
```
I can add a parameter to configure this.

---
In reply to: [555470569](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r555470569) [](ancestors = 555470569)

dmcilvaney replied:
Added new size parameter (default is 20%)

---
In reply to: [556006371](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r556006371) [](ancestors = 556006371,555470569)

MateuszMalisz commented on toolkit/tools/imager/imager.go:
```diff
@@ -464,15 +484,41 @@ func buildImage(mountPointMap, mountPointToFsTypeMap, mountPointToMountArgsMap m
+			var initramfsPathList []string
+			err = readOnlyRoot.SwitchDeviceToReadOnly(mountPointMap["/"], mountPointToMountArgsMap["/"])
+			if err != nil {
+				logger.Log.Errorf("Fail 1 %s", err.Error())
```
I think it would be a good idea to make this more descriptive.

MateuszMalisz replied:
Applies to "Fail 2" as well.

MateuszMalisz commented on toolkit/tools/internal/miscdebug/miscdebug.go:
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package debugutils
```
Shouldn't we chenge the file and folder name as well?
